### ISSUE TITLE
[Question] Calling devDependencies with CLIs after `lerna link convert`

### ISSUE THREAD

**Issue by blieque** [lerna/lerna#2020]
I was running into problems getting a new leaf package to install, and decided to move all `devDependencies` up to the root package to get around the problem. One leaf package is a Vue web app, and can no longer build as "webpack" no longer exists in node_modules/.bin. I feel like I must be doing something wrong as this seems like it would break a lot of projects.

From npm:
```
sh: webpack-dev-server: command not found
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! yourdata@0.1.0 dev: `webpack-dev-server --config build/webpack.dev.conf.js`
npm ERR! spawn ENOENT
…
```

Edit: On another note, is there a better place for questions? IRC? Discord? Gitter?

## Expected Behaviour
Using `lerna link convert` for the benefits it offers doesn't break npm scripts in leaf packages.

## Current Behaviour
It appears to do so.

## Possible Solution
Lerna symlinks /packages/<leaf package>/node_modules/.bin to /node_modules/.bin. This worked for me, so I'll create a script to perform this step in CI.

### Manual Solution
This should work if all of your leaf packages are rooted two directories below the monorepo root, as is usual. It'll probably spit out some mkdir or rm errors, but it will carry on.
```
lerna exec "rm -r node_modules/.bin ; mkdir node_modules ; ln -s ../../../node_modules/.bin node_modules/.bin"
```

## Steps to Reproduce (for bugs)
1. Install webpack-cli as a devDependency of a leaf package.
2. Run `lerna link convert`. This will remove the devDependency from the leaf package and add it to the root package.
3. Attempt to use webpack in a leaf package script.

<details><summary>lerna.json</summary><p>

```json
{
  "packages": [
    "packages/*",
    "sites/*"
  ],
  "version": "0.1.0"
}
```
</p></details>

## Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.10.7 |
| `npm --version`  | 6.9.0 |
| `yarn --version` | n/a |
| `node --version` | 11.6.0 |

| OS | Version |
| --- | --- |
| macOS High Sierra | 10.13.6 |

**Comment by ford04** [comment]
I also had the same thoughts on this topic (still beeing quite new to lerna). 

Afaik the workflow of `lerna link convert` is to pull all devDependencies up to the root and _remove_ them from the local packages. That also implies, that all CLI scripts have to be invoked in root from now on.

If you think about it, it would be inconsistent for the packages to rely on an implicit CLI script that is not existent anymore in their `package.json` `devDependencies` and therefore local `node_modules/.bin` folder. Node resolves `require` of modules by searching in local `node_modules` and if not found going up the file tree, like [here](https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders). But that does not account for CLI/bin scripts that are only searched in local `node_modules/.bin` and the global `.bin`. `npx` behaves in a similar way, so I think, that lernas way is consistent with node .

If I understand [the docs](https://github.com/lerna/lerna/tree/master/commands/bootstrap#lernabootstrap) correctly, `lerna bootstrap --hoist` would preserve the `devDependencies` in the local `package.json` and link the bins locally. So you could still execute scripts from the local packages. 

@evocateur is there an option for `lerna bootstrap --hoist --scope <mypkg>` to only hoist the `devDependencies`, when the version is the same in all packages, so you don't get a duplicate with a different version locally? Current behavior seems to be that only a [warning is shown](https://github.com/lerna/lerna/tree/master/commands/bootstrap#--hoist-glob), and not aborted with an error.
Also: Is there a filter option for `lerna link convert` to include only specific devDependencies?

**Comment by evocateur** [comment]
No. Running `lerna link convert` removes `devDependencies` in leaf nodes because they literally _do not work_ once you install those local `file:` packages from the root. You should be able to run `webpack --config ./packages/foo/webpack.dev.config.js` from the root no problem.

**Comment by clark-stevenson** [comment]
Forgive this generic question on top as I am new to monorepos.  I imported my `A` package, that package has a npm start command field set as this: `"start": "webpack-dev-server --config ./webpack.dev.js`  script. 

In the past, previous to `lerna link convert`.  I would open up `packages/A` in an editor, and run `npm start`.  This is no longer the case as explained above. 

Should my outlook when using mono repos, be composed at the root instead?  The `start` or `build` moves from inside the package, to the root in general?

- `"start:A": "webpack-dev-server --config ./packages/A/webpack.dev.js"`

**Comment by clark-stevenson** [comment]
I wrote a quick question here but removed it to go and find help elsewhere but madness is setting in!  If anyone has a second, I would appreciate it.  I am on this issue, and ultimately what is killing me is once I have done the `lerna link convert` command (after importing a project).   Obviously all my dependencies have moved into `/package.json` from `/packages/x/package.json`. 

And just like evocateur suggests, this is all good.   For example without change, `lerna run build` builds all my bits and pieces! Amazing!

There is just one niggle left for me,  in day to day usage,  I would expect that if I am going to work on project/x  I would open it in a VSCode workspace, and hit `npm start` within this context but that is obviously not the case given from what I read here. 

None the less, I still want that behaviour,  and searching for it leads me to articles which seem to do what I am after. 

For example this link: http://samhogy.co.uk/2018/04/moving-to-a-monorepo.html 

> Example 2: Similar to above, two packages depend on the same dependency, but require it to be available from the command-line. An example is Webpack, as you may wish to run Webpack from the package’s folder. When you bootstrap, Webpack will be installed at the root-level node_modules, although a “symlink” exists in the .bin folder of each package’s node_modules

```
node_modules
 \ webpack
packages
 \ package-a
  \ node_modules
   \ .bin
    \ webpack.cmd
 \ package-b
  \ node_modules
   \ .bin
    \ webpack.cmd
```


That is what I want!  But I am not seeing any evidence that it is possible or how to achieve it.  

tl;dr

Everything is grand, but I want to open up a command line in `packages/x` and type `npm start` and it works.  Today I get `webpack-dev-server cannot be found`. 

Thanks

**Comment by blieque** [comment]
@clark-stevenson I've been where you are. I have a script on the root package which will give you the behaviour you want. It's not the cleanest solution, but it does work. If your leaf packages have production dependencies, though, it may cause problems.

```json
"scripts": {
  ...
  "symlink-bins": "lerna exec \"rm -r node_modules/.bin ; mkdir node_modules ; ln -s ../../../node_modules/.bin node_modules/.bin\""
},
```

`npm run symlink-bins` in the root will then set up symlinks for webpack-dev-server and other CLIs.

**Comment by evocateur** [comment]
This is actually something `pnpm recursive` handles out of the box. Considering switching myself given the likelihood that workspaces aren't coming to `npm` any time soon.

**Comment by clark-stevenson** [comment]
Thanks a lot guys for your insight!   👍 

Before I came to lerna, I was looking at microsoft rush and I was introduced to pnmp there although I thought it was a bit "out there".  Mind you I still think Yarn is "out there" so hardly the best of judges although I might look more into this while I am on the subject.  

Appreciate it
